<a href="https://colab.research.google.com/github/nagamani0604/Nagamani_INFO5731_Fall2024/blob/main/Somireddy_Nagamani_Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment One**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [29]:

import math

def calculate_the_volume_of_hemisphere(r):
    return (2/3) * math.pi * r**3

def print_hemisphere_volumes():
    print(f"{'r':<5} {'V':<10}")
    print("-" * 15)

    for r in range(2, 21, 2):
        V = calculate_the_volume_of_hemisphere(r)
        print(f"{r:<5} {V:<10.2f}")

print_hemisphere_volumes()


r     V         
---------------
2     16.76     
4     134.04    
6     452.39    
8     1072.33   
10    2094.40   
12    3619.11   
14    5747.02   
16    8578.64   
18    12214.51  
20    16755.16  


# Question 2 (20 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [34]:
# Write code for each of the sub parts with proper comments.

import random

# (a) Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.
random_numbers = [random.randint(50, 150) for _ in range(12)]
print("(A) Random Numbers:", random_numbers)


# (b) Generated numbers in ascending order
ascending_sorted_numbers = sorted(random_numbers)
print("(B) Generated numbers in ascending order:", ascending_sorted_numbers)

# (c) Program to arrange the numbers in descending order
descending_sorted_numbers = sorted(random_numbers, reverse=True)
print("(C) Generated numbers in descending order:", descending_sorted_numbers)

# (d) Average of the entire group of generated numbers
average_all = sum(random_numbers) / len(random_numbers)
print(f"(D) Average of entire generated numbers: {average_all:.2f}")

# (e) Average of the first 6 numbers from both the ascending and descending sorted lists.
average_first6_ascending = sum(ascending_sorted_numbers[:6]) / 6
average_first6_descending = sum(descending_sorted_numbers[:6]) / 6

print(f"(E) The average of first 6 numbers from ascending list: {average_first6_ascending:.2f}")
print(f"The average of first 6 numbers from descending list): {average_first6_descending:.2f}")









(A) Random Numbers: [58, 130, 130, 111, 135, 96, 146, 59, 143, 102, 97, 65]
(B) Generated numbers in ascending order: [58, 59, 65, 96, 97, 102, 111, 130, 130, 135, 143, 146]
(C) Generated numbers in descending order: [146, 143, 135, 130, 130, 111, 102, 97, 96, 65, 59, 58]
(D) Average of entire generated numbers: 106.00
(E) The average of first 6 numbers from ascending list: 79.50
The average of first 6 numbers from descending list): 132.50


# Question 3 (20 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [35]:

import string

# Remove punctuation
def remove_punctuation(sentence):
    return sentence.translate(str.maketrans('', '', string.punctuation))

# Count vowels in a sentence
def count_vowels(sentence):
    vowels = "aeiou"
    sentence = sentence.lower()
    vowel_count = {vowel: sentence.count(vowel) for vowel in vowels}
    return vowel_count

# a. Input for test
sentence1 = "Hello , how are you doing today ?"
sentence2 = "My name is Mani and she is my friend Abhi . We are going to the same school"

# b. Sentence transformation
sentence1_list = sentence1.split()
sentence2_list = sentence2.split()
combined_list = sentence1_list + sentence2_list

# Output
print("\nCombined list of words: ")
print(combined_list)

# c: Compare and find common words
common_words = set(sentence1_list) & set(sentence2_list)
print("\nCommon words between both sentences: ")
print(list(common_words))

# d Remove punctuation from the sentences
sanitized_sentence1 = remove_punctuation(sentence1)
sanitized_sentence2 = remove_punctuation(sentence2)
print("\nSanitized Sentence 1: ")
print(sanitized_sentence1)
print("\nSanitized Sentence 2: ")
print(sanitized_sentence2)

# e Number of vowels in each sentence
vowels_count_sentence1 = count_vowels(sanitized_sentence1)
vowels_count_sentence2 = count_vowels(sanitized_sentence2)

print("Sentence 1: ")
print(vowels_count_sentence1)
print("Sentence 2: ")
print(vowels_count_sentence2)




Combined list of words: 
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'My', 'name', 'is', 'Mani', 'and', 'she', 'is', 'my', 'friend', 'Abhi', '.', 'We', 'are', 'going', 'to', 'the', 'same', 'school']

Common words between both sentences: 
['are']

Sanitized Sentence 1: 
Hello  how are you doing today 

Sanitized Sentence 2: 
My name is Mani and she is my friend Abhi  We are going to the same school
Sentence 1: 
{'a': 2, 'e': 2, 'i': 1, 'o': 5, 'u': 1}
Sentence 2: 
{'a': 6, 'e': 7, 'i': 6, 'o': 4, 'u': 0}


# Question 4 (20 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [27]:
# Your code here

import math

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error:  Cannot be divided by zero"
    except Exception as e:
        return f"Error: Invalid input - {str(e)}"

def calculator():
    print("Regular Calculator. Type 'exit' to quit.")
    while True:
        expression = input("Please enter an expression: ")
        if expression.lower() == 'exit':
            print("Exiting calculator")
            break
        result = evaluate_expression(expression)
        print(f"Result: {result}")

calculator()


def evaluate_expression(expression):
    try:
        result = eval(expression, {"sqrt": math.sqrt, "pow": math.pow, "__builtins__": None}, {})
        return result
    except ZeroDivisionError:
        return "Error: Cannot be divided by zero"
    except Exception as e:
        return f"Error: Invalid input - {str(e)}"

def calculator():
    print("Extended Calculator with Exponentiation, Square Root, and Modulus. Type 'exit' to quit.")
    print("Supported operations: +, -, *, /, ** (exponentiation), % (modulus), sqrt(number)")
    while True:
        expression = input("Please enter an expression: ")
        if expression.lower() == 'exit':
            print("Exiting calculator.")
            break
        result = evaluate_expression(expression)
        print(f"Result: {result}")

calculator()




Regular Calculator. Type 'exit' to quit.
Please enter an expression: 2+2
Result: 4
Please enter an expression: 2-2
Result: 0
Please enter an expression: 3*3
Result: 9
Please enter an expression: 4/3
Result: 1.3333333333333333
Please enter an expression: exit
Exiting calculator
Extended Calculator with Exponentiation, Square Root, and Modulus. Type 'exit' to quit.
Supported operations: +, -, *, /, ** (exponentiation), % (modulus), sqrt(number)
Please enter an expression: sqrt(169)
Result: 13.0
Please enter an expression: 7/0
Result: Error: Cannot be divided by zero
Please enter an expression: 0/9
Result: 0.0
Please enter an expression: exit
Exiting calculator.


# Question 5 (30 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (10 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (10 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [36]:
import pandas as pd

# Load the dataset
file_path = '/content/sample_data/Assignment1_denton_housing.csv'
df = pd.read_csv(file_path)

# Check the structure of the dataset
df.head()

# a Calculate the difference in the number of Occupied Housing Units from year to year and print
occupied_df = df[df['title_field'] == 'Occupied Housing Units'].copy()
occupied_df = occupied_df.sort_values(by='year')
occupied_df['yearly_difference'] = occupied_df['value'].diff()

occupied_diff_sorted = occupied_df[['year', 'yearly_difference']].dropna().sort_values(by='yearly_difference')

# Results
print("(A) Number of Occupied Housing Units from year to year")
print(occupied_diff_sorted)



 # b Calculate the percentage of housing units which are vacant an occupied
occupied_df = df[df['title_field'] == 'Occupied Housing Units']
vacant_df = df[df['title_field'] == 'Vacant Housing Units']
total_df = df[df['title_field'] == 'Total Housing Units']

merged_df = pd.merge(occupied_df[['year', 'value']], vacant_df[['year', 'value']], on='year', suffixes=('_occupied', '_vacant'))
merged_df = pd.merge(merged_df, total_df[['year', 'value']], on='year')
merged_df.rename(columns={'value': 'total_units'}, inplace=True)

merged_df['Occupied Housing Units'] = (merged_df['value_occupied'] / merged_df['total_units']) * 100
merged_df['Vacant Housing Units'] = (merged_df['value_vacant'] / merged_df['total_units']) * 100

percentage_df = merged_df[['year', 'Vacant Housing Units', 'Occupied Housing Units']].sort_values(by='year')

# Results
print("(B)", percentage_df.to_string(index=False))


# c the highest number of housing units were vacant and occupied
vacant_df = df[df['title_field'] == 'Vacant Housing Units']
occupied_df = df[df['title_field'] == 'Occupied Housing Units']

max_vacant = vacant_df.loc[vacant_df['value'].idxmax()]
max_occupied = occupied_df.loc[occupied_df['value'].idxmax()]

max_values_df = pd.DataFrame({
    '': ['Vacant Housing Units', 'Occupied Housing Units'],
    'Year': [max_vacant['year'], max_occupied['year']],
    'Value': [max_vacant['value'], max_occupied['value']]
})

# Result
print("(C) Highest Vacant and Occupied Housing Units:")
print(max_values_df.to_string(index=False))




# d Harmonic mean of the total housing units
from scipy.stats import hmean

total_units_df = df[df['title_field'] == 'Total Housing Units'].copy()

harmonic_mean_total_units = hmean(total_units_df['value'])

print("(D) Harmonic mean of the total housing units is" , harmonic_mean_total_units )


(A) Number of Occupied Housing Units from year to year
    year  yearly_difference
17  2009             -795.0
11  2011             -303.0
8   2012             1404.0
5   2013             1565.0
2   2014             2622.0
14  2010             5091.0
(B)  year  Vacant Housing Units  Occupied Housing Units
 2008             10.882653               89.117347
 2009              9.071116               90.928884
 2010              8.142556               91.857444
 2011              9.379522               90.620478
 2012              6.677600               93.322400
 2013              8.801788               91.198212
 2014              5.730111               94.269889
(C) Highest Vacant and Occupied Housing Units:
                        Year  Value
  Vacant Housing Units  2008   4483
Occupied Housing Units  2014  46295
(D) Harmonic mean of the total housing units is 44398.28543969185


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [38]:
# Write your response below

Working with the random numbers is one of the challenging part since we have to work and validate the code withunknown inputs. Also working with the data set where we calculate percentages,
differences is also challenging for me. Though it was challenging i enjoyed working with the random numbers. Since i was  maths student earlier i also enjoyed the calculator where i had
implemented all the mathematical expressions.


SyntaxError: invalid syntax (<ipython-input-38-0d07c0e6bee0>, line 3)